File: coding_skill_cpp/19_mutex_guard.cpp
/* C++ Mutex Guard - lock_guard */
#include <iostream>
#include <mutex>
using namespace std;

mutex mtx;

void critical_section() {
    lock_guard<mutex> guard(mtx);  // 자동 lock
    cout << "[Critical] 보호된 영역" << endl;
}  // 자동 unlock

int main() {
    cout << "=== C++ Mutex Guard ===" << endl;
    critical_section();
    cout << "자동 unlock 완료!" << endl;
    return 0;
}

File: coding_skill_py/19_mutex_guard.py
"""Python Mutex Guard - with Lock()"""
from threading import Lock

lock = Lock()

def critical_section():
    with lock:  # 자동 acquire
        print("[Critical] 보호된 영역")
    # 자동 release

def main():
    print("=== Python Mutex Guard ===")
    critical_section()
    print("자동 release 완료!")

if __name__ == "__main__":
    main()
